### PR TITLE
Fix #13947 - Get type parameters for a TypeRepr

### LINKED ISSUE #13947: No way to list the type parameters from a quotes Symbol or TypeRepr
It's not possible in quotes to do something like:

```scala
def foo(s: Symbol): List[Symbol] = s.typeParams
```

In dotc it's possible to:

```scala
def foo(s: Symbol)(using Context): List[TypeSymbol] = s.denot.typeParams
```

### PR DESCRIPTION
**Pull request by cchantep**

### CODE REVIEW

**Comment by KacperFKorban** [comment]
Mima filters need to be updated. Necessary filters are printed in logs and should be added to `project/MimaFilters.scala`.

**Comment by cchantep** [comment]
> Mima filters need to be updated. Necessary filters are printed in logs and should be added to `project/MimaFilters.scala`.

Done

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -2571,6 +2571,9 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
         @experimental
         def substituteTypes(from: List[Symbol], to: List[TypeRepr]): TypeRepr
 
+        /** The type parameters */
```
Comment (nicolasstucki) [inline library/src/scala/quoted/Quotes.scala:2574]: Should mention that these are the arguments of a type application and also what happens if it is not a type application.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -1749,6 +1749,10 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
           dotc.core.Types.decorateTypeApplications(self).appliedTo(targs)
         def substituteTypes(from: List[Symbol], to: List[TypeRepr]): TypeRepr =
           self.subst(from, to)
+
+        def typeParams: List[TypeRepr] = self match
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:1753]: This is getting the arguments and not the parameters, I would call it `typeArgs`.

Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:1753]: `typeParams` provides references to the definitions of the parameters. Here we provide the type arguments passed to the type constructor.

Comment (cchantep) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: Not convinced by the naming difference, but not strong opinion.

File: library/src/scala/quoted/Quotes.scala [hunk 3]
```diff
@@ -2571,6 +2571,9 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
         @experimental
         def substituteTypes(from: List[Symbol], to: List[TypeRepr]): TypeRepr
 
+        /** The applied type parameters (empty if there is no such parameters) */
```
Comment (nicolasstucki) [inline library/src/scala/quoted/Quotes.scala:2574]:
```suggestion
        /** The applied type arguments (empty if there is no such arguments) */
```

File: project/MiMaFilters.scala [hunk 4]
```diff
@@ -9,6 +9,8 @@ object MiMaFilters {
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.runtime.Tuples.append"),
     ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.substituteTypes"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.substituteTypes"),
+    ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.typeParams"),
```
Comment (nicolasstucki) [inline project/MiMaFilters.scala:12]:
```suggestion
    ProblemFilters.exclude[ReversedMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.typeArgs"),
```

File: project/MiMaFilters.scala [hunk 5]
```diff
@@ -9,6 +9,8 @@ object MiMaFilters {
...
+    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.typeParams"),
```
Comment (nicolasstucki) [inline project/MiMaFilters.scala:13]:
```suggestion
    ProblemFilters.exclude[DirectMissingMethodProblem]("scala.quoted.Quotes#reflectModule#TypeReprMethods.typeArgs"),
```

File: tests/run-macros/i13947/Macro_1.scala [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import scala.quoted.*
+
+inline def printTypeParams[A]: Unit = ${ printTypeParamsImpl[A] }
+
+def printTypeParamsImpl[A: Type](using q: Quotes): Expr[Unit] = {
+  import q.reflect.*
```
Comment (nicolasstucki) [inline tests/run-macros/i13947/Macro_1.scala:6]:
```suggestion
  import quotes.reflect.*
```

File: tests/run-macros/i13947/Macro_1.scala [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import scala.quoted.*
+
+inline def printTypeParams[A]: Unit = ${ printTypeParamsImpl[A] }
+
+def printTypeParamsImpl[A: Type](using q: Quotes): Expr[Unit] = {
```
Comment (nicolasstucki) [inline tests/run-macros/i13947/Macro_1.scala:5]:
```suggestion
def printTypeParamsImpl[A: Type](using Quotes): Expr[Unit] = {
```

**Comment by nicolasstucki** [comment]
@cchantep could you also squash the commits and rebase?

File: tests/run-macros/i13947/Macro_1.scala (hunk 6 above)
Comment (cchantep) [reply tests/run-macros/i13947/Macro_1.scala:6]: Not convinced, `quotes` can more easily be misunderstood as `scala.quotes`

**Comment by nicolasstucki** [comment]
This was merged in #14088

**Pull request closed without merging**